source/_posts/2024-08-19-state-of-generics-and-collections.md

```diff
@@ -0,0 +1,449 @@
+---
+title: 'State of Generics and Collections'
+layout: post
+tags:
+  - news
+author:
+  - name: Arnaud Le Blanc
+    url: https://github.com/arnaud-lb
+  - name: Derick Rethans
+    url: https://derickrethans.nl
+  - name: Larry Garfield
+    url: https://www.garfieldtech.com/
+published_at: 19 August 2024
+---
+
+Generics have been on the list of wanted features for a long time by numerous PHP developers. The topic is often brought up in "What's New in PHP?" talks as well during Q&A.
+
+In this article we will be exploring the different approaches, and what their current state is.
+
+- [Full Reified generics](#full-reified-generics)
+- [Collections](#collections)
+- [Other alternatives](#other-alternatives)
+  - [Static Analysis](#static-analysis)
+  - [Erased Generic Type Declarations](#erased-generic-type-declarations)
+  - [Fully Erased Type Declarations](#fully-erased-type-declarations)
+- [Generic Arrays](#generic-arrays)
+  - [Fluid Arrays](#fluid-arrays)
+  - [Static Arrays](#static-arrays)
+- [Conclusion](#conclusion)
+
+# Full Reified generics
+
+With generics you can define classes with placeholders for the types of their properties and methods. These can then be specified when instances of the class are created. This enables code reusability and type safety across different data types.  "Reified" generics are implementations where the generic type information is defined and carried through to runtime, allowing for runtime enforcement of generic requirements.
+
+As PHP syntax, this could look like:
+
+	class Entry<KeyType, ValueType>
+	{
+		public function __construct(protected KeyType $key, protected ValueType $value)
+		{
+		}
+
+		public function getKey(): KeyType
+		{
+			return $this->key;
+		}
+
+		public function getValue(): ValueType
+		{
+			return $this->value;
+		}
+	}
+
+	new Entry<int, BlogPost>("123", new BlogPost());
```

Comment:
@derickr @pronskiy Maybe this meant to be an integer?

``` suggestion
	new Entry<int, BlogPost>(123, new BlogPost());
```
---

Thank you so much for this article!.. clear, great depth and eye-opening 🤍🙏